File: proxy/src/main/java/one/hyro/proxy/commands/BanCommand.java
package one.hyro.proxy.commands;

import com.mojang.brigadier.Command;
import com.mojang.brigadier.arguments.StringArgumentType;
import com.mojang.brigadier.tree.LiteralCommandNode;
import com.velocitypowered.api.command.BrigadierCommand;
import com.velocitypowered.api.command.CommandSource;
import com.velocitypowered.api.command.VelocityBrigadierMessage;
import com.velocitypowered.api.proxy.ProxyServer;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;

import java.util.Map;

public class BanCommand {
    public static BrigadierCommand createBanCommand(final ProxyServer proxy) {
        LiteralCommandNode<CommandSource> node = BrigadierCommand.literalArgumentBuilder("ban")
                .requires(source -> source.hasPermission("hyro.ban"))
                .executes(context -> {
                    CommandSource source = context.getSource();
                    source.sendMessage(Component.translatable("info.errors.providePLayer", NamedTextColor.RED));
                    return Command.SINGLE_SUCCESS;
                })
                .then(BrigadierCommand.requiredArgumentBuilder("target", StringArgumentType.word())
                        .suggests((ctx, builder) -> {
                            proxy.getAllPlayers().forEach(player -> builder.suggest(
                                    player.getUsername(),
                                    VelocityBrigadierMessage.tooltip(Component.text(player.getUsername()))
                            ));
                            return builder.buildFuture();
                        })
                        .executes(context -> {
                            CommandSource source = context.getSource();
                            source.sendMessage(Component.translatable("info.errors.provideReason", NamedTextColor.RED));
                            return Command.SINGLE_SUCCESS;
                        })
                        .then(BrigadierCommand.requiredArgumentBuilder("reason", StringArgumentType.word())
                                .suggests((ctx, builder) -> {
                                    Map<String, String> reasons = Map.of(
                                            "Hacks", "info.reasons.hacks",
                                            "Spamming", "info.reasons.spam",
                                            "Offensive", "info.reasons.offensive",
                                            "Other", "info.reasons.other",
                                            "None", "info.reasons.none"
                                    );

                                    for (Map.Entry<String, String> entry : reasons.entrySet()) {
                                        builder.suggest(
                                                entry.getKey(),
                                                VelocityBrigadierMessage.tooltip(Component.text(entry.getKey()))
                                        );
                                    }

                                    return builder.buildFuture();
                                })
                                .executes(context -> {
                                    CommandSource source = context.getSource();
                                    String reasonArgument = context.getArgument("reason", String.class);
                                    String targetArgument = context.getArgument("target", String.class);

                                    Map<String, String> reasons = Map.of(
                                            "Hacks", "info.reasons.hacks",
                                            "Spamming", "info.reasons.spam",
                                            "Offensive", "info.reasons.offensive",
                                            "Other", "info.reasons.other",
                                            "None", "info.reasons.none"
                                    );

                                    if (!reasons.containsKey(reasonArgument)) {
                                        source.sendMessage(Component.translatable("info.errors.invalidReason", NamedTextColor.RED));
                                        return Command.SINGLE_SUCCESS;
                                    }

                                    disconnectPLayerAndSaveBan(proxy, source, targetArgument, reasons.get(reasonArgument));
                                    return Command.SINGLE_SUCCESS;
                                })
                        )
                )
                .build();

        return new BrigadierCommand(node);
    }

    private static void disconnectPLayerAndSaveBan(final ProxyServer proxy, CommandSource source, String... arguments) {
        proxy.getPlayer(arguments[0])
                .ifPresentOrElse(player -> {
                            player.disconnect(Component.translatable("info.moderation.ban", NamedTextColor.RED));

                            // Save ban to database

                            Component successMessage = Component.translatable(
                                    "commands.ban.success",
                                    Component.text(arguments[0], NamedTextColor.DARK_GREEN)
                            ).color(NamedTextColor.GREEN);
                            source.sendMessage(successMessage);
                        },
                        () -> source.sendMessage(Component.translatable("info.errors.playerNotFound", NamedTextColor.RED))
                );
    }
}

File: proxy/src/main/kotlin/commands/BanCommand.kt
package one.hyro.commands

import com.mojang.brigadier.arguments.StringArgumentType
import com.mojang.brigadier.tree.LiteralCommandNode
import com.velocitypowered.api.command.BrigadierCommand
import com.velocitypowered.api.command.CommandSource
import com.velocitypowered.api.proxy.Player
import com.velocitypowered.api.proxy.ProxyServer
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor

object BanCommand {
    private val banReasons = mapOf(
        "Hacking" to "moderation.reasons.hacking",
        "Inappropriate" to "moderation.reasons.inappropriate",
        "Other" to "moderation.reasons.other",
        "Spamming" to "moderation.reasons.spamming",
        "Toxicity" to "moderation.reasons.toxicity"
    )

    fun createBanCommand(proxyServer: ProxyServer): BrigadierCommand {
        val banCommandNode: LiteralCommandNode<CommandSource> = BrigadierCommand.literalArgumentBuilder("ban")
            .requires { source -> source.hasPermission("hyro.ban") }
            .executes { context ->
                val commandSource: CommandSource = context.source
                commandSource.sendMessage(Component.translatable("context.error.specifyPlayer", NamedTextColor.RED))
                return@executes 0
            }
            .then(BrigadierCommand.requiredArgumentBuilder("player", StringArgumentType.word())
                .suggests { _, builder ->
                    val onlinePlayers = proxyServer.allPlayers.map { it.username }
                    onlinePlayers.forEach { builder.suggest(it) }
                    builder.buildFuture()
                }
                .executes { context ->
                    val commandSource: CommandSource = context.source
                    commandSource.sendMessage(Component.translatable("context.error.specifyReason", NamedTextColor.RED))
                    return@executes 1
                }
                .then(BrigadierCommand.requiredArgumentBuilder("reason", StringArgumentType.word())
                    .suggests { _, builder ->
                        banReasons.forEach { (reason, _) -> builder.suggest(reason) }
                        builder.buildFuture()
                    }
                    .executes { context ->
                        val commandSource: CommandSource = context.source
                        val playerName = StringArgumentType.getString(context, "player")
                        val targetPlayer = proxyServer.getPlayer(playerName)
                        val reason = StringArgumentType.getString(context, "reason")

                        if (reason !in banReasons.keys) {
                            commandSource.sendMessage(Component.translatable("context.errors.invalidReason", NamedTextColor.RED))
                            return@executes 1
                        }

                        if (targetPlayer.isPresent) banPlayer(targetPlayer.get(), commandSource, reason)
                        else commandSource.sendMessage(Component.translatable("context.error.playerOffline", NamedTextColor.RED))
                        return@executes 1
                    }
                )
            )
            .build()

        return BrigadierCommand(banCommandNode)
    }

    private fun banPlayer(player: Player, commandSource: CommandSource, reason: String) {
        // TODO: Add ban to database
        player.disconnect(Component.translatable("context.success.banMessage", NamedTextColor.RED))
        val successMessage = Component.translatable(
            "context.success.ban",
            Component.text(player.username, NamedTextColor.DARK_GREEN)
        ).color(NamedTextColor.GREEN)
        commandSource.sendMessage(successMessage)
    }
}